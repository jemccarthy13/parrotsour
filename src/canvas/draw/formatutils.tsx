/**
 * This file contains utilities for group and answer formatting
 */
import { AircraftGroup } from "../../classes/groups/group"
import { toRadians } from "../../utils/psmath"
import { ArrowDataTrail } from "../../classes/aircraft/datatrail/arrowdatatrail"

/**
 * Convert an altitude to a 3-digit flight level
 * @param alt int altitude to change to padded string
 */
export function formatAlt(alt: number): string {
  return (alt * 10).toString().substring(0, 3).padStart(3, "0")
}

/**
 * Check if picture is opening or closing
 * @param fg First group of picture
 * @param sg Second group of picture
 */
export function getOpenCloseAzimuth(
  fg: AircraftGroup,
  sg: AircraftGroup
): string {
  let retVal = ""

  const fgStartPos = fg.getStartPos()
  const sgStartPos = sg.getStartPos()

  const angle1to2 = fgStartPos.getBR(sgStartPos).bearingNum

  const fgHeading = fg.getHeading()
  const sgHeading = sg.getHeading()

  const speed1to2 =
    ArrowDataTrail.LEN_TRAIL * Math.cos(toRadians(angle1to2 - fgHeading))

  const speed2to1 =
    ArrowDataTrail.LEN_TRAIL * Math.cos(toRadians(angle1to2 - sgHeading))

  const acceptableRate = Math.cos(toRadians(5))

  let closing = speed1to2 > 0 && speed2to1 < 0
  let opening = speed1to2 < 0 && speed2to1 > 0

  closing =
    closing ||
    (Math.abs(speed1to2) < acceptableRate && speed2to1 < 0) ||
    (Math.abs(speed2to1) < acceptableRate && speed1to2 > 0)
  opening =
    opening ||
    (Math.abs(speed1to2) < acceptableRate && speed2to1 > 0) ||
    (Math.abs(speed2to1) < acceptableRate && speed1to2 < 0)

  const parallel =
    Math.abs(speed1to2) < acceptableRate && Math.abs(speed2to1) < acceptableRate

  if (!parallel) {
    if (opening) {
      retVal = " OPENING "
    }
    if (closing) {
      retVal = " CLOSING "
    }
  }

  return retVal
}
