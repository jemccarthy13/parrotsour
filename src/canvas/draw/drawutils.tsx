/**
 * This file contains common low-level drawing utilities used
 * to build groups and pictures.
 */

// Interfaces
import { Point } from "../../classes/point"

/**
 * 'Clamp' the location to the confines of the drawing context
 * @param ctx The context to constrict it to
 * @param pos the current position
 */
export function clampInContext(
  ctx: CanvasRenderingContext2D,
  x: number | Point,
  y?: number
): Point {
  let retPoint = Point.DEFAULT
  if (x instanceof Point) {
    retPoint = new Point(
      Math.min(Math.max(x.x, 0), ctx.canvas.width),
      Math.min(Math.max(x.y, 0), ctx.canvas.height)
    )
  } else {
    if (y === undefined) y = 0
    retPoint = new Point(
      Math.min(Math.max(x, 0), ctx.canvas.width),
      Math.min(Math.max(y, 0), ctx.canvas.height)
    )
  }
  return retPoint
}
