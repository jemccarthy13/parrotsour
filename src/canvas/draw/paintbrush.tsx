import { Braaseye } from "../../classes/braaseye"
import { AircraftGroup } from "../../classes/groups/group"
import { Point } from "../../classes/point"
import { PIXELS_TO_NM, randomNumber } from "../../utils/psmath"
import { PictureCanvasProps, PictureCanvasState } from "../canvastypes"
import { clampInContext } from "./drawutils"
import { formatAlt } from "./formatutils"

export class PaintBrush {
  private static ctx: CanvasRenderingContext2D

  constructor(ctx: CanvasRenderingContext2D) {
    PaintBrush.ctx = ctx
  }

  public static use(ctx: CanvasRenderingContext2D | undefined | null): void {
    if (ctx) PaintBrush.ctx = ctx
  }

  public static getContext(): CanvasRenderingContext2D {
    return PaintBrush.ctx
  }

  public static clearCanvas(): void {
    const ctx = PaintBrush.ctx
    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height)
    ctx.fillStyle = "white"
    ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height)
  }

  /**
   * Draw a line on the drawing context, given the properties
   * @param startX start x of the line
   * @param startY start y of the line
   * @param endX end x of the line
   * @param endY end y of the line
   * @param color (optional) color of the line, defaults to black
   */
  public static drawLine(
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    color = "black",
    forcedCtx?: CanvasRenderingContext2D
  ): void {
    let context = PaintBrush.ctx
    if (forcedCtx) context = forcedCtx
    context.lineWidth = 1
    context.strokeStyle = color
    context.beginPath()
    context.moveTo(startX, startY)
    context.lineTo(endX, endY)
    context.stroke()
    context.stroke()
  }

  /**
   * Draw text to the drawing context at the given position
   * @param text Text to draw
   * @param x X Position to draw at
   * @param y Y Position to draw at
   * @param size (optional) Size of the text to draw, defaults 12
   * @param color (optional) Color of the text, defaults to black
   */
  public static drawText(
    text: string,
    x: number,
    y: number,
    size = 12,
    color = "black"
  ): void {
    const ctx = PaintBrush.ctx
    ctx.lineWidth = 1
    ctx.fillStyle = color
    ctx.font = size + "px Arial"
    const pos = clampInContext(ctx, x, y)

    ctx.fillText(text, pos.x, pos.y)
    ctx.fillText(text, pos.x, pos.y)
  }

  /**
   * Draw a measurement (distance with a length number)
   * @param {Point} start position to start for line
   * @param {Point} end position to end for line
   * @param {number} distance how long the line should be
   * @param {boolean} showMeasurements Display the measurements
   */
  public static drawMeasurement(
    start: Point,
    end: Point,
    distance: number,
    showMeasurements: boolean
  ): void {
    if (showMeasurements) {
      PaintBrush.drawLine(start.x, start.y, end.x, end.y)
      PaintBrush.drawText(
        Math.floor(distance).toString(),
        (start.x + end.x) / 2,
        (start.y + end.y) / 2 - 3
      )
    }
  }

  /**
   * Draw altitudes next to a group with optional offset
   *
   * @param grpPos group's current position
   * @param alts altitudes to draw
   * @param offX (optional) x-axis offset from group
   * @param offY (optional) y-axis offset from group
   */
  public static drawAltitudes(
    grpPos: Point,
    alts: number[],
    offX?: number,
    offY?: number
  ): void {
    const offsetX = offX || 0
    const offsetY = offY || 0
    const formattedAlts: string[] = alts.map((a: number) => {
      return formatAlt(a)
    })
    PaintBrush.drawText(
      formattedAlts.join(","),
      grpPos.x + 25 + offsetX,
      grpPos.y - 11 + offsetY,
      11,
      "#ff8c00"
    )
  }

  public static drawBullseye(bull?: Point, color?: string): Point {
    const context = PaintBrush.ctx
    color = color || "black"
    context.lineWidth = 1
    context.fillStyle = color
    context.strokeStyle = color

    const centerPointX = bull
      ? bull.x
      : randomNumber(context.canvas.width * 0.33, context.canvas.width * 0.66)
    const centerPointY = bull
      ? bull.y
      : randomNumber(context.canvas.height * 0.33, context.canvas.height * 0.66)

    context.beginPath()
    context.arc(
      centerPointX,
      centerPointY,
      PIXELS_TO_NM / 2,
      0,
      2 * Math.PI,
      true
    )
    context.stroke()
    context.fill()

    context.moveTo(centerPointX, centerPointY + PIXELS_TO_NM * 2)
    context.lineTo(centerPointX, centerPointY - PIXELS_TO_NM * 2)
    context.stroke()

    context.moveTo(centerPointX + PIXELS_TO_NM * 2, centerPointY)
    context.lineTo(centerPointX - PIXELS_TO_NM * 2, centerPointY)
    context.stroke()

    return new Point(centerPointX, centerPointY)
  }

  public static drawFullInfo(
    state: PictureCanvasState,
    props: PictureCanvasProps,
    groups: AircraftGroup[]
  ): void {
    for (let y = 0; y < groups.length; y++) {
      const grpPos = groups[y].getCenterOfMass(props.dataStyle)
      if (props.showMeasurements) {
        new Braaseye(
          grpPos,
          state.blueAir.getCenterOfMass(props.dataStyle),
          state.bullseye
        ).draw(true, props.braaFirst)
      }
      PaintBrush.drawAltitudes(grpPos, groups[y].getAltitudes())
    }
  }
}
