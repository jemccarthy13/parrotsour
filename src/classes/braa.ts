import { PaintBrush } from "../canvas/draw/paintbrush"

/**
 * BRAA is a class that contains a bearing and range
 * as well as formatting and drawing functions for any bearing/range.
 */
export class BRAA {
  bearing = "000"
  bearingNum = 0
  range = 0
  label = ""

  constructor(bearingNum: number, range: number) {
    this.bearingNum = bearingNum
    this.bearing = this.bearingNum.toString().padStart(3, "0")
    this.range = range
  }

  /**
   * Formatter
   *
   * @returns Formatted bearing and range
   */
  public toString(): string {
    return this.bearing + "/" + this.range
  }

  /**
   * Draw a bearing and range at the given coordinates.
   *
   * @param ctx Current drawing context
   * @param x X position of draw
   * @param y Y position of draw
   * @param color
   * @param showMeasurements
   */
  draw(x: number, y: number, color?: string, showMeasurements?: boolean): void {
    if (showMeasurements) {
      PaintBrush.drawText(this.toString(), x, y, 11, color)
    }
  }
}
