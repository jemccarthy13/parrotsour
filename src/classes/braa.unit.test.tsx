import { PaintBrush } from "../canvas/draw/paintbrush"
import { BRAA } from "./braa"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const canvasSerializer = require("jest-canvas-snapshot-serializer")

expect.addSnapshotSerializer(canvasSerializer)

describe("BRAA", () => {
  let canvas: HTMLCanvasElement
  let ctx: CanvasRenderingContext2D

  beforeAll(() => {
    canvas = document.createElement("canvas")
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ctx = canvas.getContext("2d")!
    canvas.height = 20
    canvas.width = 50
    PaintBrush.use(ctx)
  })

  afterEach(() => {
    ctx.fillStyle = "white"
    ctx.fillRect(0, 0, canvas.width, canvas.height)
  })

  it("constructs_correctly", () => {
    const br = new BRAA(90, 20)
    expect(br.bearing).toEqual("090")
    expect(br.range).toEqual(20)
    expect(br.bearingNum).toEqual(90)
  })

  it("converts_to_pretty_string", () => {
    const br = new BRAA(90, 20)
    expect(br.toString()).toEqual("090/20")
  })

  it("draws_correctly", () => {
    const br = new BRAA(90, 20)
    br.draw(10, 10, "black", true)
    expect(canvas).toMatchSnapshot()
  })

  it("draws_only_when_showmeasure_true", () => {
    const br = new BRAA(90, 20)
    br.draw(10, 10, "black", false)
    expect(canvas).toMatchSnapshot()
  })
})
