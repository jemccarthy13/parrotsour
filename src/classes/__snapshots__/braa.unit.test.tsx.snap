// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BRAA draws_correctly 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAUCAYAAADPym6aAAAABmJLR0QA/wD/AP+gvaeTAAAB4ElEQVRIie2VQctpURSG177nHHSQiRIDGUj5F4oJGVLmfoKJgRFlJnPMlH9gYIyZksyFEUknGRg+d3Bz+nRzuPdTVzdP7cnufdfab3uvtgKQ/4Af//oAr+LpIN1uV9xutyilxOv1SqvVEhGR4XAoPp9PlFISiURks9k46kVEJpOJGIYhvV5PTNMUpZSYpin9fv+h9y48STgcplAosNvtSKfThEIhAPx+P5VKBcuyqNfrJJNJRz1AuVwmHo/jcrmo1WpYlkWpVMLj8Tz03uPpIJqmsV6vAZjNZhiG8auACLvdDoD9fo+maY56gHg8Trlcvql/OBxQSj30fjuIiHC5XAA4nU52U9M0aTQaWJZFsVjkesn39AC6rjMej2/qN5tN+zadvN8OopTidDoBYFmWXXwwGODxeFBKkclk7P17+vF4jK7rN7UXiwXBYJDtduvofUkQwzBYLpcAzOdz3G73b5rRaGS/83v663xcmc/nRKNRptPpH/X66yCxWIx8Ps/xeCSXyxEOh+2m1WqV1WpFNpsllUo56r/OR6PRIBAI2Id+1OslQdrtNpqmISK4XC46nY69bxgGSikSiQTn89lR/3U+dF1HRG6Wk9cJBZ+f/a34BHk3PkHejU+Qd+MT5N34CXAvYcJf4ALbAAAAAElFTkSuQmCC"
  height="20"
  width="50"
/>
`;

exports[`BRAA draws_only_when_showmeasure_true 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAUCAYAAADPym6aAAAABmJLR0QA/wD/AP+gvaeTAAAANklEQVRIie3PgQkAMAjEwNr9d36XEAySmyCpJHkH/O2AKY7QOELjCI0jNI7QOELjCI0jNI7QNNSvBCS3oAPJAAAAAElFTkSuQmCC"
  height="20"
  width="50"
/>
`;
